nilearn/decomposition/base.py

```diff
@@ -0,0 +1,535 @@
+        provided ratio. Default: 'auto'
+
+    n_components: integer, optional
+        Number of components to be extracted by the PCA
+
+    random_state: int or RandomState
+        Pseudo number generator state used for random sampling.
+
+    memory_level: integer, optional
+        Integer indicating the level of memorization. The higher, the more
+        function calls are cached.
+
+    memory: joblib.Memory
+        Used to cache the function calls.
+
+    memory_mode: {'auto', 'array', 'memorymap'}
+        Whether to return a memorymap or and ndarray. Default: 'auto'
+
+    Retuns
+    ------
+    data: ndarray or memorymap
+        Concatenation of reduced data
+    """
+
+    def __init__(self, masker, imgs, confounds=None,
+                 reduction_ratio='auto',
+                 n_components=None, random_state=None,
+                 memory_level=0,
+                 memory=Memory(cachedir=None),
+                 max_nbytes=1e9):
+        self.masker = masker
+        self.imgs = imgs
+        self.confounds = confounds
+        self.reduction_ratio = reduction_ratio
+        self.n_components = n_components
+        self.random_state = random_state
+        self.memory_level = memory_level
+        self.memory = memory
+        self.max_nbytes = max_nbytes
+
+    def __enter__(self):
+        if isinstance(self.imgs, nibabel.Nifti1Image):
+            imgs = [self.imgs]
+        else:
+            imgs = self.imgs
+
+        if self.reduction_ratio != 'auto':
+            reduction_ratio = float(self.reduction_ratio)
+            if reduction_ratio is None or reduction_ratio >= 1:
+                reduction_ratio = 1
+        else:
+            if self.n_components is None:
+                reduction_ratio = 1
+            else:
+                reduction_ratio = 'n_components'
+        # Precomputing number of samples for preallocation
+        subject_n_samples = np.zeros(len(imgs), dtype='int')
+
+        for i, img in enumerate(imgs):
+            if reduction_ratio == 'n_components':
+                subject_n_samples[i] = min(self.n_components,
+                                           check_niimg_4d(img).shape[3])
+            else:
+                subject_n_samples[i] = int(ceil(check_niimg_4d(img).
+                                           shape[3] * reduction_ratio))
+        subject_limits = np.zeros(subject_n_samples.shape[0]+1,
+                                  dtype='int')
+        subject_limits[1:] = np.cumsum(subject_n_samples)
+        n_voxels = np.sum(self.masker.mask_img_.get_data())
+        n_samples = subject_limits[-1]
+
+        if self.max_nbytes is not None:
+            max_nbytes = int(self.max_nbytes)
+            return_mmap = n_voxels * n_samples * 8 > max_nbytes
+        else:
+            return_mmap = False
+
+        # We initialize data in memory or on disk
+        if return_mmap:
+            temp_folder = _get_dataset_dir('temp', verbose=0)
```

Comment:
Again, write access is not guaranteed here. Is there any reason to save the file here instead of the system temporary folder?